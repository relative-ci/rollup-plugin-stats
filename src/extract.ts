import { OutputAsset, OutputBundle, OutputChunk, RenderedModule } from 'rollup';
import omit from 'lodash/omit.js';

export type AssetStats = Omit<OutputAsset, 'source'> & {
  source?: OutputAsset['source'];
};

export type ModuleStats = Omit<RenderedModule, 'code'> & {
  code?: RenderedModule['code'] | null;
}

export type ChunkStats = Omit<OutputChunk, 'code' | 'modules'> & {
  code?: OutputChunk['code'];
  modules: Record<string, ModuleStats>;
};

export type Stats = Record<string, AssetStats | ChunkStats>;

export type StatsOptions = {
  /**
   * Output asset/module sources
   * @default false 
   */
  source?: boolean;
}

export default function extractRollupStats(bundle: OutputBundle, options: StatsOptions = {}): Stats {
  const { source = false } = options;

  const output: Stats = {};

  Object.entries(bundle).forEach(([key, entry]) => {
    if (entry.type === "asset") {
      let entryAsset = structuredClone(entry) as AssetStats;

      // Skip asset source if options.source is false
      if (!source) {
        entryAsset = omit(entryAsset, 'source');
      }

      output[key] = entryAsset;

      return;
    }

    if (entry.type === "chunk") {
      let entryChunk = structuredClone(entry) as ChunkStats;

      // Skip chunk code if options.source is false
      if (!source) {
        entryChunk = omit(entryChunk, 'code');
      }

      Object.entries(entryChunk.modules).forEach(([moduleKey, moduleEntry]) => {
        let entryChunkModule = structuredClone(moduleEntry) as ModuleStats;

        // Skip module source if source is false
        if (!source) {
          entryChunkModule = omit(entryChunkModule, 'code');
        }

        entryChunk.modules[moduleKey] = entryChunkModule;
      });

      output[key] = entryChunk;

      return;
    }
  });
  
  return output;
}
