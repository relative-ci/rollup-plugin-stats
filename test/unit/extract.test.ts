import { describe, test, expect } from 'vitest';

import extract from '../../src/extract';
import * as rollupStats from './fixtures/rollup-stats';

describe('extract', () => {
  test('should extract rollup stats', () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    expect(extract(rollupStats.stats as any)).toMatchSnapshot();
  });

  test('should extract rollup stats with sources', () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    expect(extract(rollupStats.stats as any, { source: true })).toMatchSnapshot();
  });
});
