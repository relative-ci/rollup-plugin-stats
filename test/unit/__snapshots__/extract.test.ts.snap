// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extract > should extract rollup stats 1`] = `
{
  "assets/logo.abcd1234.svg": {
    "fileName": "assets/logo.abcd1234.svg",
    "type": "asset",
  },
  "assets/main.abcd1234.js": {
    "fileName": "assets/main.abcd1234.js",
    "modules": {
      "/home/user/project/src/main.js": {},
      "/home/user/project/src/utils.js": {},
    },
    "type": "chunk",
  },
  "assets/vendors.abcd1234.js": {
    "fileName": "assets/vendors.abcd1234.js",
    "modules": {
      "/home/user/project/node_modules/package-a/index.js": {},
      "/home/user/project/node_modules/package-b/index.js": {},
    },
    "type": "chunk",
  },
}
`;

exports[`extract > should extract rollup stats with excluded assets 1`] = `
{
  "assets/logo.abcd1234.svg": {
    "fileName": "assets/logo.abcd1234.svg",
    "type": "asset",
  },
  "assets/main.abcd1234.js": {
    "fileName": "assets/main.abcd1234.js",
    "modules": {
      "/home/user/project/src/main.js": {},
      "/home/user/project/src/utils.js": {},
    },
    "type": "chunk",
  },
}
`;

exports[`extract > should extract rollup stats with excluded modules 1`] = `
{
  "assets/logo.abcd1234.svg": {
    "fileName": "assets/logo.abcd1234.svg",
    "type": "asset",
  },
  "assets/main.abcd1234.js": {
    "fileName": "assets/main.abcd1234.js",
    "modules": {
      "/home/user/project/src/main.js": {},
    },
    "type": "chunk",
  },
  "assets/vendors.abcd1234.js": {
    "fileName": "assets/vendors.abcd1234.js",
    "modules": {
      "/home/user/project/node_modules/package-a/index.js": {},
      "/home/user/project/node_modules/package-b/index.js": {},
    },
    "type": "chunk",
  },
}
`;

exports[`extract > should extract rollup stats with sources 1`] = `
{
  "assets/logo.abcd1234.svg": {
    "fileName": "assets/logo.abcd1234.svg",
    "source": "<svg></svg>",
    "type": "asset",
  },
  "assets/main.abcd1234.js": {
    "code": "export default function () {}",
    "fileName": "assets/main.abcd1234.js",
    "modules": {
      "/home/user/project/src/main.js": {
        "code": "export default function () {}",
      },
      "/home/user/project/src/utils.js": {
        "code": "export default function () {}",
      },
    },
    "type": "chunk",
  },
  "assets/vendors.abcd1234.js": {
    "code": "export default function () {}",
    "fileName": "assets/vendors.abcd1234.js",
    "modules": {
      "/home/user/project/node_modules/package-a/index.js": {
        "code": "export default function () {}",
      },
      "/home/user/project/node_modules/package-b/index.js": {
        "code": "export default function () {}",
      },
    },
    "type": "chunk",
  },
}
`;
